lib/Transforms/SecretInsertMgmt/SecretInsertMgmtCKKS.cpp

```diff
@@ -184,28 +209,51 @@ struct SecretInsertMgmtCKKS
                                std::move(patternsBootstrapWaterLine));
 
     // when other binary op operands level mismatch
-    // includeFirstMul not used for these ops
+    int scaleCounter = -1;  // for making adjust_scale op different to avoid cse
     RewritePatternSet patternsAddModReduce(&getContext());
-    patternsAddModReduce.add<ModReduceBefore<arith::AddIOp>>(
-        &getContext(), /*isMul*/ false, /*includeFirstMul*/ false,
-        getOperation(), &solver);
-    patternsAddModReduce.add<ModReduceBefore<arith::AddFOp>>(
-        &getContext(), /*isMul*/ false, /*includeFirstMul*/ false,
-        getOperation(), &solver);
-    patternsAddModReduce.add<ModReduceBefore<arith::SubIOp>>(
-        &getContext(), /*isMul*/ false, /*includeFirstMul*/ false,
-        getOperation(), &solver);
-    patternsAddModReduce.add<ModReduceBefore<arith::SubFOp>>(
-        &getContext(), /*isMul*/ false, /*includeFirstMul*/ false,
-        getOperation(), &solver);
+    patternsAddModReduce.add<MatchCrossLevel<arith::AddIOp>>(
+        &getContext(), &scaleCounter, getOperation(), &solver);
+    patternsAddModReduce.add<MatchCrossLevel<arith::SubIOp>>(
+        &getContext(), &scaleCounter, getOperation(), &solver);
+    patternsAddModReduce.add<MatchCrossLevel<arith::MulIOp>>(
+        &getContext(), &scaleCounter, getOperation(), &solver);
+    patternsAddModReduce.add<MatchCrossLevel<arith::AddFOp>>(
+        &getContext(), &scaleCounter, getOperation(), &solver);
+    patternsAddModReduce.add<MatchCrossLevel<arith::SubFOp>>(
+        &getContext(), &scaleCounter, getOperation(), &solver);
+    patternsAddModReduce.add<MatchCrossLevel<arith::MulFOp>>(
+        &getContext(), &scaleCounter, getOperation(), &solver);
     (void)walkAndApplyPatterns(getOperation(), std::move(patternsAddModReduce));
 
+    // when other binary op operands mulDepth mismatch
+    // this only happen for before-mul but not include-first-mul case
+    // at the first level, a Value can be both mulResult or not mulResult
+    // we should match their scale by adding one adjust scale op
+    if (!beforeMulIncludeFirstMul && !afterMul) {
+      RewritePatternSet patternsMulDepth(&getContext());
+      patternsMulDepth.add<MatchCrossMulDepth<arith::MulIOp>>(
+          &getContext(), &scaleCounter, getOperation(), &solver);
+      patternsMulDepth.add<MatchCrossMulDepth<arith::AddIOp>>(
+          &getContext(), &scaleCounter, getOperation(), &solver);
+      patternsMulDepth.add<MatchCrossMulDepth<arith::SubIOp>>(
+          &getContext(), &scaleCounter, getOperation(), &solver);
+      patternsMulDepth.add<MatchCrossMulDepth<arith::MulFOp>>(
+          &getContext(), &scaleCounter, getOperation(), &solver);
+      patternsMulDepth.add<MatchCrossMulDepth<arith::AddFOp>>(
+          &getContext(), &scaleCounter, getOperation(), &solver);
+      patternsMulDepth.add<MatchCrossMulDepth<arith::SubFOp>>(
+          &getContext(), &scaleCounter, getOperation(), &solver);
```

Comment:
```suggestion
      patternsMulDepth.add<MatchCrossMulDepth<arith::MulIOp>, MatchCrossMulDepth<arith::AddIOp>, MatchCrossMulDepth<arith::SubIOp>, MatchCrossMulDepth<arith::MulFOp>, MatchCrossMulDepth<arith::AddFOp>, MatchCrossMulDepth<arith::SubFOp>>(
          &getContext(), &scaleCounter, getOperation(), &solver);
```

Since they all take the same constructor arguments, I suspect you can use the variable template args version of `RewritePatternSet::add`, here and elsewhere